src/backend/cuda/nvrtc/cache.cpp

```diff
@@ -457,21 +457,73 @@ Kernel buildKernel(const int device, const string &nameExpr,
     return entry;
 }
 
+Kernel loadKernel(const int device, const string &nameExpr) {
+    const string &cacheDirectory = getKernelCacheDirectory();
+    if (cacheDirectory.empty()) return Kernel{nullptr, nullptr};
+
+    const string cacheFile = cacheDirectory + AF_PATH_SEPARATOR
+        + getKernelCacheFilename(device, nameExpr);
+
+    CUmodule module = nullptr;
+    CUfunction kernel = nullptr;
+
+    try
+    {
+        std::ifstream in(cacheFile, std::ios::binary);
+        if (!in.is_open()) return Kernel{ nullptr, nullptr };
+
+        in.exceptions(std::ios::failbit | std::ios::badbit);
+
+        size_t nameSize;
+        in >> nameSize;
+        string name;
+        name.resize(nameSize);
+        in.read(&name[0], nameSize);
+
+        size_t cubinSize;
+        in >> cubinSize;
+        vector<char> cubin(cubinSize);
+        in.read(cubin.data(), cubinSize);
+        in.close();
+
+        CU_CHECK(cuModuleLoadDataEx(&module, cubin.data(), 0, 0, 0));
+        CU_CHECK(cuModuleGetFunction(&kernel, module, name.c_str()));
+
+        AF_TRACE("{{{:<30} : loaded from {} for {} }}",
+            nameExpr, cacheFile, getDeviceProp(device).name);
+    } catch (...) {
+        if (module != nullptr) {
+            CU_CHECK(cuModuleUnload(module));
+        }
+        removeFile(cacheFile);
+        throw;
```

Comment:
It should be `AF_ERROR("message", AF_ERR_RUNTIME)` instead of an empty throw.

Reply:
See comment above; maybe we don't want to throw from this function and just return an empty kernel, so that we fallback to building the kernel again if loading fails.